### ISSUE TITLE
TypeScript mixin extends other mixin: inconsistent “conflicting declarations” error

### ISSUE THREAD

**Issue by cspotcode** [Microsoft/TypeScript#22845]
**TypeScript Version:**   typescript@2.9.0-dev.20180323

**Search Terms:** 

mixin extends mixin
mixin conflicting declarations
mixin base contraint

**Summary**

I want to write two mixin classes, `Foo` and `Bar`. `Bar` requires access to the protected interface of `Foo`. Therefore I use TS2.8 conditional types to extract the type of a `Foo` class and use this type as a generic constraint for the `Bar` mixin's `Base` class.

Unfortunately, this causes mysteriously inconsistent "conflicting declaration" errors.  The inconsistency suggests this is a bug instead of an intended type warning.

**Code**

```ts
class BaseClass {/* ... */}

/* Type error highlighted on "Composed" */
class Composed extends BarMixin(FooMixin(BaseClass)) {}

type Constructor<I = {}> = new (...args: any[]) => I;

interface EmptyInterface {}

/** Mixin */
function FooMixin<C extends Constructor>(Base: C) {
    return class extends Base {
        private _referenceToEmptyInterface: EmptyInterface = {}; // Causes the error.  Why?
        private _fooPrivate: number = 0; // Does not cause an error.  Why not, given the above?
        protected _fooProtected: number = 0;
    }
}

/** Type of class that creates instances of type Foo */
type FooConstructor = typeof FooMixin extends (a: Constructor) => infer Cls ? Cls : never;

/** Mixin that can only extend subclasses of Foo */
function BarMixin<C extends FooConstructor>(Base: C) {
    return class extends Base {
        barMethod() {
            // We require `Base` to implement `Foo` because we need access to protected `Foo` properties.
            this._fooProtected++;
        }
    }
}
```

**Expected behavior:**

No errors.

**Actual behavior:**

```
index.ts(4,7): error TS2415: Class 'Composed' incorrectly extends base class 'BarMixin<{ new (...args: any[]): FooMixin<typeof BaseClass>.(Anonymous class); prototype: FooMixi...'.
  Type 'Composed' is not assignable to type 'BarMixin<{ new (...args: any[]): FooMixin<typeof BaseClass>.(Anonymous class); prototype: FooMixi...'.
    Property '_referenceToEmptyInterface' has conflicting declarations and is inaccessible in type 'Composed'.
```

[**Playground Link** ](http://www.typescriptlang.org/play/#src=class%20BaseClass%20%7B%2F*%20...%20*%2F%7D%0D%0A%0D%0A%2F*%20Type%20error%20highlighted%20on%20%22Composed%22%20*%2F%0D%0Aclass%20Composed%20extends%20BarMixin(FooMixin(BaseClass))%20%7B%7D%0D%0A%0D%0Atype%20Constructor%3CI%20%3D%20%7B%7D%3E%20%3D%20new%20(...args%3A%20any%5B%5D)%20%3D%3E%20I%3B%0D%0A%0D%0Ainterface%20EmptyInterface%20%7B%7D%0D%0A%0D%0A%2F**%20Mixin%20*%2F%0D%0Afunction%20FooMixin%3CC%20extends%20Constructor%3E(Base%3A%20C)%20%7B%0D%0A%20%20%20%20return%20class%20extends%20Base%20%7B%0D%0A%20%20%20%20%20%20%20%20private%20_referenceToEmptyInterface%3A%20EmptyInterface%20%3D%20%7B%7D%3B%20%2F%2F%20Causes%20the%20error.%20%20Why%3F%0D%0A%20%20%20%20%20%20%20%20private%20_fooPrivate%3A%20number%20%3D%200%3B%20%2F%2F%20Does%20not%20cause%20an%20error.%20%20Why%20not%2C%20given%20the%20above%3F%0D%0A%20%20%20%20%20%20%20%20protected%20_fooProtected%3A%20number%20%3D%200%3B%0D%0A%20%20%20%20%7D%0D%0A%7D%0D%0A%0D%0A%2F**%20Type%20of%20class%20that%20creates%20instances%20of%20type%20Foo%20*%2F%0D%0Atype%20FooConstructor%20%3D%20typeof%20FooMixin%20extends%20(a%3A%20Constructor)%20%3D%3E%20infer%20Cls%20%3F%20Cls%20%3A%20never%3B%0D%0A%0D%0A%2F**%20Mixin%20that%20can%20only%20extend%20subclasses%20of%20Foo%20*%2F%0D%0Afunction%20BarMixin%3CC%20extends%20FooConstructor%3E(Base%3A%20C)%20%7B%0D%0A%20%20%20%20return%20class%20extends%20Base%20%7B%0D%0A%20%20%20%20%20%20%20%20barMethod()%20%7B%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20%2F%2F%20We%20require%20%60Base%60%20to%20implement%20%60Foo%60%20because%20we%20need%20access%20to%20protected%20%60Foo%60%20properties.%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20this._fooProtected%2B%2B%3B%0D%0A%20%20%20%20%20%20%20%20%7D%0D%0A%20%20%20%20%7D%0D%0A%7D)  (will not work until the playground is updated to TS2.8)

**Related Issues:**

I asked on StackOverflow: https://stackoverflow.com/questions/49458031/

**Comment by andy-ms** [comment]
Simplified repro:
```ts
type Constructor = new (...args: any[]) => {};

const Mixin1 = <C extends Constructor>(Base: C) => class extends Base { private _fooPrivate: {}; }

type FooConstructor = typeof Mixin1 extends (a: Constructor) => infer Cls ? Cls : never;
const Mixin2 = <C extends FooConstructor>(Base: C) => class extends Base {};

/* Type error highlighted on "Composed" */
class Composed extends Mixin2(Mixin1(Object)) {}
```
